### ISSUE TITLE
Library folders are not recognized when editing remote files

### ISSUE THREAD

**Issue by ztlpn** [emacs-lsp/lsp-mode#1069]
**Describe the bug**
I am editing files on a remote machine using Tramp and I want to jump to the definition of some library function. This works but lsp-mode doesn't recognize the opened file as a library file and suggests to add it as part of a new project (`vec_deque.rs is not part of any project. Select action:`)

**To Reproduce**
1. Open a file on a remote machine with Tramp
2. Jump to the definiton of a library function.

**Expected behavior**
The `library-folders-fn` method returns a list `("~/.cargo/registry/src" "~/.rustup/toolchains")`. As I jump to the file `/ssh:user@host:/home/user/.rustup/toolchains/beta-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/liballoc/collections/vec_deque.rs` I expect it to be recognized as a library file.

**Which Language Server did you use**
RLS

**OS**
macOS (remote OS Linux)

I guess the problematic code is [here](https://github.com/emacs-lsp/lsp-mode/blob/228326bfead89dd9bcd508603d15e1b12f711851/lsp-mode.el#L5799) but my emacs-fu is insufficient to fix it.

**Comment by yyoncho** [comment]
You could fix it in your tramp rls registration. Just specify a function that returns the list of folders that are libraries on the remote machine like here: https://github.com/emacs-lsp/lsp-mode/blob/master/lsp-rust.el#L298

Let me know if you will need a help with that.

**Comment by yyoncho** [comment]
Also, you might try to add `/ssh:user@host:/home/user/.rustup/` to `lsp-rust-library-directories`.

**Comment by ztlpn** [comment]
Well that would mean that I need to list all the hosts I connect to in the `lsp-rust-library-directories` variable... Workable but a bit ugly. I guess tramp-mode has spoiled me and I expect a better degree of transparency :)

I guess there are really two problems here:
1. the library file path includes the remote hostname
2. the home directory on the remote host is different than on the local host

I think we can at least solve problem 1. by substituting `(buffer-file-name)` with `(lsp-file-local-name (buffer-file-name))` [here](https://github.com/emacs-lsp/lsp-mode/blob/228326bfead89dd9bcd508603d15e1b12f711851/lsp-mode.el#L5799), is that right?

**Comment by yyoncho** [comment]
> Workable but a bit ugly. 

That is true, but given the fact that we have very limited resources, it is impossible to polish every workflow. I wish lsp-mode was able to detect that it is on the remote server, automatically download and install the servers if they are not present, preconfigure the language paths and so on...

> I think we can at least solve problem 1. by substituting (buffer-file-name) with (lsp-file-local-name (buffer-file-name)) here, is that right?

I would first check how expand-file-name works when you are on the remote machine. In case it works as expected we may change the code to detect that the original file is on a remote machine and expand the files so they match the remote one. We may also extend lsp-mode initial import menu with option - add to library paths.

**Comment by ztlpn** [comment]
> That is true, but given the fact that we have very limited resources, it is impossible to polish every workflow.

I didn't mean to complain, lsp-mode is awesome. Just trying to make it a little better :)

> I would first check how expand-file-name works when you are on the remote machine. In case it works as expected we may change the code to detect that the original file is on a remote machine and expand the files so they match the remote one. 

As the path in question (`~/.rustup/toolchains`) is absolute, `expand-file-name` expands as if it were local (i.e. to `/Users/local_user/.rustup/toolchains`) - which kind of makes sense. If I stick the server location in front of the path (i.e. `/ssh:user@host:~/.rustup/toolchains`) then it gets expanded properly (to `/ssh:user@host:/home/user/.rustup/toolchains`). So I guess the correct algorithm should be something like:
1. check if the buffer filename is remote, if it is, extract the server location from it.
2. check if the library folder is local, if it is, stick the server location in front of it.
3. expand the resulting library folder so that `f-is-ancestor-of?` works correctly.
